File: src/main/java/komu/blunt/core/CoreExtractTagExpression.java
package komu.blunt.core;

import komu.blunt.analyzer.VariableReference;
import komu.blunt.asm.Assembler;
import komu.blunt.asm.Instructions;
import komu.blunt.asm.Linkage;
import komu.blunt.asm.Register;

import static com.google.common.base.Preconditions.checkNotNull;

public class CoreExtractTagExpression extends CoreExpression {

    private final PatternPath path;
    private final VariableReference var;

    public CoreExtractTagExpression(VariableReference var, PatternPath path) {
        this.var = checkNotNull(var);
        this.path = checkNotNull(path);
    }

    @Override
    public void assemble(Assembler asm, Instructions instructions, Register target, Linkage linkage) {
        instructions.loadVariable(target, var);
        instructions.loadTag(target, target, path);
        instructions.finishWithLinkage(linkage);
    }

    @Override
    public String toString() {
        return "(extract-tag " + var.name + " " + path + ")";
    }
}

File: src/main/kotlin/komu/blunt/core/CoreExtractTagExpression.kt
package komu.blunt.core

import komu.blunt.analyzer.VariableReference
import komu.blunt.asm.*

class CoreExtractTagExpression(private val variable: VariableReference, private val path: PatternPath) : CoreExpression() {

    override fun assemble(asm: Assembler, target: Register, linkage: Linkage) =
        instructions {
            loadVariable(target, variable)
            loadTag(target, target, path)
            finishWithLinkage(linkage)
        }

    override fun simplify() = this
    override fun toString() = "(extract-tag ${variable.name} $path)"
}
